Implement Key Metrics Selection Panel

nfmohit opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #6258 

## Relevant technical choices

This PR implements the Key Metrics Selection Panel.

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

aaemnnosttv commented on assets/sass/components/global/_googlesitekit-side-sheet.scss:
```diff
@@ -0,0 +1,60 @@
+		top: 0;
+		transition: right 1s;
+		width: 100%;
+		z-index: 10002;
```
Please use add a new variable to the list which was just added 👍 Also double check that the value makes sense relative to the others.
https://github.com/google/site-kit-wp/blob/c0c351575827676bb49b2a0b9c26274632f4d183/assets/sass/config/_variables.scss#L315-L329

nfmohit replied:
Applied, thank you!

nfmohit commented on assets/js/googlesitekit/components-gm2/Checkbox.js:
```diff
@@ -45,15 +45,26 @@ export default function Checkbox( props ) {
 		onKeyDown,
 		loading,
 		alignLeft,
+		onClick,
 	} = props;
 
+	const extraProps = {};
+
+	if ( onClick ) {
+		extraProps.onClick = onClick;
+		extraProps.onKeyDown = () => {};
+		extraProps.tabIndex = 0;
+		extraProps.role = 'button';
+	}
+
 	return (
 		<div className="mdc-form-field">
 			<div
 				className={ classnames( 'mdc-checkbox', {
 					'mdc-checkbox--disabled': disabled,
 					'mdc-checkbox--align-left': alignLeft,
 				} ) }
+				{ ...extraProps }
```
Optionally allowing `onClick` on `div.mdc-checkbox`. In our scenario, we're doing this to prevent opening the metric accordion on checkbox click. See:

https://github.com/google/site-kit-wp/blob/cb05c080fccc1655021bfb6f2f89a22e7df761ac/assets/js/components/KeyMetrics/MetricsSelectionPanel/Metrics.js#L100-L109

techanvil replied:
Hey @nfmohit, if making changes to the `Checkbox` (or any component which has a GM3 counterpart when we have more of them), please can you either make a corresponding change to the GM3 version to keep it in sync, or create a followup issue to do so?

aaemnnosttv commented on assets/js/googlesitekit/datastore/user/key-metrics.js:
```diff
@@ -305,6 +332,18 @@ const baseSelectors = {
+	 * @param {Object} state Data store's state.
+	 * @return {boolean} TRUE if the key metrics settings are being saved, otherwise FALSE.
+	 */
+	isSavingKeyMetricsSettings( state ) {
+		return !! state?.isSavingKeyMetricsSettings;
+	},
```
We should be able to do this without the additional saving state. We've needed to do the same a few times before. Here's one example https://github.com/google/site-kit-wp/blob/2dffea28d9e449cab27159189f59e77285ba3a93/assets/js/modules/analytics/datastore/profiles.js#L257-L261

aaemnnosttv commented on assets/sass/components/global/_googlesitekit-accordion.scss:
```diff
@@ -80,6 +80,18 @@
+		.googlesitekit-accordion__header {
+			pointer-events: none;
+
+			&::after {
+				// This is the expand_more icon (arrow/chevron down), in the color $c-interactive-on-disable-container.
+				background-image: url("data:image/svg+xml;charset=UTF-8,%3Csvg%20width%3D%2212%22%20height%3D%228%22%20viewBox%3D%220%200%2012%208%22%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%3E%3Cpath%20d%3D%22M10.6.6L12%202%206%208%200%202%201.4.6%206%205.2z%22%20fill%3D%22%23030E07%22%20fill-opacity%3D%220.32%22%20fill-rule%3D%22evenodd%22%2F%3E%3C%2Fsvg%3E");
+			}
```
Perhaps one for another issue but this should ideally use the icon as an svg/component rather than encoded like this where we have to override it every time for changing the color or something else.

nfmohit replied:
I agree. Opened #7363 for a wider change across the codebase. Let me know if it looks good, thanks!

aaemnnosttv commented on assets/sass/components/global/_googlesitekit-side-sheet.scss:
```diff
@@ -0,0 +1,59 @@
+	}
+
+	.googlesitekit-side-sheet-overlay {
+		background-color: rgba(0, 0, 0, 0.6);
```
Do we not have a variable for this?

nfmohit replied:
Doesn't look like we have a variable for this specific one, but I've created one now. Let me know if that looks good, thanks!

aaemnnosttv commented on assets/sass/components/global/_googlesitekit-side-sheet.scss:
```diff
@@ -0,0 +1,59 @@
+		position: fixed;
+		top: 0;
+		width: 100%;
+		z-index: $z-index-sk-side-sheet - 10; // -10 for the side sheet to appear above the overlay.
```
While it probably seems a bit silly to do so, let's track this with its own dedicated variable with the others. Part of the idea there is that we catalog all of the indexes we use in one place to avoid unintentionally positioning something in the wrong plane.

nfmohit replied:
Updated, thanks!

aaemnnosttv commented on assets/sass/config/_variables.scss:
```diff
@@ -327,3 +327,4 @@ $max-width-settings-notice-content: 1098px;
 
 // Ensures header does not interfere with important wp dashboard menu elements.
 $z-index-sk-header: 9980;
+$z-index-sk-side-sheet: 10000;
```
Please list in descending order.

nfmohit replied:
Updated, thanks!

aaemnnosttv commented on assets/js/components/KeyMetrics/MetricsSelectionPanel/Footer.js:
```diff
@@ -0,0 +1,168 @@
+					{
+						br: <br />,
+						link: (
+							<Link
+								onClick={ onSettingsClick }
+								disabled={ isSavingSettings }
+							/>
```
This link uses a different color than used in the design `#108080`

aaemnnosttv reviewed (changes requested):
Thanks @nfmohit ! This is a lot to do in one PR. Even though I left more comments than usual, it doesn't seem like that much relative to the overall size 😄 Great work!

nfmohit commented on assets/sass/components/global/_googlesitekit-side-sheet.scss:
```diff
@@ -0,0 +1,59 @@
+		right: 0;
+		top: 0;
+		transform: translateX(100%);
+		transition: transform 1s;
```
I can do that, however, just to confirm, the dashboard sharing modal/Material dialogs have a transition duration of 75ms:

![image](https://github.com/google/site-kit-wp/assets/20284937/972cd749-f420-4ffa-bf03-69b93336294d)

Here's how it looks with the Side Sheet:

https://github.com/google/site-kit-wp/assets/20284937/c2c8c53a-687a-4cd6-bb79-d2a935d6c3d6

Do you think it is a bit too quick?

Let me know what you think, thanks!

nfmohit replied:
As discussed in our 1:1, I've reduced it to 0.5 seconds. Thanks!

nfmohit commented on assets/js/components/ErrorText.js:
```diff
@@ -32,16 +32,25 @@ import { sprintf, __ } from '@wordpress/i18n';
 	if ( ! message ) {
 		return null;
 	}
 
-	let error = sprintf(
-		/* translators: %s: Error message */
-		__( 'Error: %s', 'google-site-kit' ),
-		message
-	);
+	let error = message;
+
+	if ( prefix ) {
+		error = sprintf(
+			/* translators: 1: Prefix, 2: Error message */
+			__( '%1$s %2$s', 'google-site-kit' ),
+			prefix,
+			message
+		);
+	}
```
Makes absolute sense. I've updated it to use `noPrefix`. Thank you!

nfmohit commented on assets/js/components/KeyMetrics/MetricsSelectionPanel/Metrics.js:
```diff
@@ -0,0 +1,123 @@
+				const id = `key-metric-selection-checkbox-${ metric }`;
+
+				return (
+					<div
+						key={ id }
+						className="googlesitekit-km-selection-panel-metrics__metric-item"
+					>
+						<Accordion
+							title={
+								<Checkbox
+									checked={ selectedMetrics?.includes(
+										metric
+									) }
+									onChange={ ( event ) => {
+										onMetricCheckboxChange( event, metric );
+									} }
+									onClick={ ( event ) => {
+										event.stopPropagation();
+									} }
+									disabled={
+										! selectedMetrics?.includes( metric ) &&
+										selectedMetrics?.length > 3
+									}
+									id={ id }
+									name={ id }
+									value={ metric }
+								>
+									{ title }
+								</Checkbox>
+							}
+							disabled={
+								! selectedMetrics?.includes( metric ) &&
+								selectedMetrics?.length > 3
+							}
+						>
+							{ description }
+						</Accordion>
+					</div>
```
Great idea, done, thanks!

nfmohit commented on assets/js/components/KeyMetrics/MetricsSelectionPanel/Footer.js:
```diff
@@ -0,0 +1,168 @@
+			[ KEY_METRICS_SELECTED ]: keyMetrics,
+		} );
+
+		setValue( KEY_METRICS_SELECTION_PANEL_OPENED_KEY, false );
```
Updated to support escape key and clicking elsewhere to close SideSheet, let me know what you think, thanks!

nfmohit commented on assets/js/components/KeyMetrics/MetricsSelectionPanel/index.js:
```diff
@@ -0,0 +1,68 @@
+
+	const { setValues } = useDispatch( CORE_FORMS );
+
+	useEffect( () => {
+		setValues( KEY_METRICS_SELECTION_FORM, {
+			[ KEY_METRICS_SELECTED ]: keyMetrics || [],
+		} );
+	}, [ keyMetrics, setValues ] );
```
Updated to reset only on SideSheet open. Let me know what you think, thanks!

nfmohit commented on assets/js/components/SideSheet.js:
```diff
@@ -0,0 +1,69 @@
+export default function SideSheet( { className, children, isOpen } ) {
+	// Disable scrolling on document body when panel is open.
+	useEffect( () => {
+		if ( isOpen ) {
+			document.body.style.overflow = 'hidden';
+		} else {
+			document.body.style.removeProperty( 'overflow' );
```
Updated, thank you!

nfmohit commented on assets/js/components/KeyMetrics/MetricsSelectionPanel/index.test.js:
```diff
@@ -0,0 +1,185 @@
+	} );
+
+	describe( 'Metrics', () => {
+		it( 'should only list metrics dependant on connected modules', () => {
```
Updated, thank you!

nfmohit commented on assets/js/googlesitekit/datastore/user/key-metrics.js:
```diff
@@ -102,26 +105,44 @@ const baseActions = {
 	 */
-	*saveKeyMetricsSettings() {
+	*saveKeyMetricsSettings( settings = {} ) {
+		invariant( isPlainObject( settings ), 'Settings should be an object.' );
```
Updated, thank you!

nfmohit commented on assets/js/components/SideSheet.stories.js:
```diff
@@ -0,0 +1,60 @@
+export default {
+	title: 'Components/SideSheet',
+	component: SideSheet,
+	decorators: [
+		( Story ) => {
+			return <Story />;
+		},
+	],
```
Whoops, my mistake, removed, thank you!

nfmohit commented on assets/js/components/SideSheet.js:
```diff
@@ -0,0 +1,69 @@
+				className={ classnames( 'googlesitekit-side-sheet', className, {
+					'googlesitekit-side-sheet--open': isOpen,
+				} ) }
+			>
```
How about we use [`aria-modal` with a role of `dialog`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-modal)? Even though this isn't technically a modal, it does act like one, at least to screen readers. What do you think?

aaemnnosttv replied:
Yeah, I'm not sure which is most appropriate (or both?) but it seems like some indication is needed.

tofumatt replied:
Both is the way to go for this one, especially as it'll be grabbing focus when opened, right?

When hidden we should do `aria-hidden`, when visible the `aria-modal="true"` + `role="dialog"` works. It's functionally a modal I'd say, so I think it's fine to label it as one.

👍🏻

nfmohit replied:
Applied accordingly, thank you!

Just a note that following the dashboard sharing settings dialog, I've left `aria-modal="true"` + `role="dialog" present both when it is hidden and visible, but adjusted `aria-hidden` to be `true` only when visible. Let me know if I should do it otherwise.

Thank you!

nfmohit commented:
> Thanks @nfmohit ! This is a lot to do in one PR. Even though I left more comments than usual, it doesn't seem like that much relative to the overall size 😄 Great work!

Thank you for the very kind and detailed review, @aaemnnosttv. I've addressed the feedback, and left a few questions. Please let me know how this looks now, thank you!

aaemnnosttv commented on assets/js/components/Accordion.js:
```diff
@@ -44,12 +45,26 @@ export default function Accordion( {
 		}
 	}, [ isActive, onClose, onOpen ] );
 
+	useEffect( () => {
+		if ( disabled && isActive ) {
+			setActive( false );
+		}
+	}, [ disabled, isActive ] );
```
Ah good point. This isn't great because it creates a layout shift where the item under the mouse changes.

My first thought is that the behavior of the accordion only toggles the visibility of the description which doesn't necessarily need to be disabled by the selection limits (only the checkboxes really need to be disabled). It could be useful to be able to see what the descriptions are without deselecting one to be able to read it.

Let's restore the previous behavior to avoid the layout shift and address this in a follow-up as we probably need to consult with UX.

nfmohit replied:
I agree. The accordion doesn't need to be disabled in any case. I've restored the previous behavior and will open a follow-up issue to discuss about the accordion. Thanks!

aaemnnosttv replied:
@nfmohit Do we have a new issue for the follow-up?

nfmohit replied:
Just opened #7383 for the follow-up. Apologies for the delay!

aaemnnosttv commented on assets/js/components/Accordion.js:
```diff
@@ -44,12 +45,26 @@ export default function Accordion( {
+
 		setActive( ! isActive );
-	}, [ isActive ] );
+	}, [ disabled, isActive ] );
```
Good thinking, but you're right that we can't rely on the browser to guard against important rules. A lot of things are possible via dev tools that aren't via the interface but that's okay so long as the backend does its job 👍

nfmohit replied:
Understood, thanks!

aaemnnosttv commented on assets/sass/components/key-metrics/_googlesitekit-km-selection-panel.scss:
```diff
@@ -0,0 +1,132 @@
+		}
+	}
+
+	.googlesitekit-km-selection-panel-metrics {
+		overflow: auto;
+
+		.googlesitekit-km-selection-panel-metrics__metric-item {
+			margin-bottom: $grid-gap-phone;
+			padding: 0 $grid-gap-desktop;
+
+			.googlesitekit-accordion {
+				border: 1px solid $c-neutral-n-66;
+				border-radius: 8px;
+				padding: 0 20px;
+
+				.googlesitekit-accordion__header {
```
One of the principles of BEM is to minimize nesting by making individual selectors more specific. Let's apply that throughout to minimize the specificity we're using to avoid creating problems in the future when we need to override something.

Practically speaking, everything here should be nested under the top level plugin class, but below that we should try to flatten it out as much as we can while still only targeting what we need to.

Classes like `__metric-item` shouldn't be nested under its parent, same with accordion and it's `__header`.

nfmohit replied:
Updated, thank you!

aaemnnosttv commented on assets/sass/components/key-metrics/_googlesitekit-km-selection-panel.scss:
```diff
@@ -0,0 +1,132 @@
+				.googlesitekit-accordion__header {
+					padding: 12px 0;
+
+					.mdc-checkbox {
+						margin-left: -11px;
```
Are we missing this class in the implementation? https://github.com/google/site-kit-wp/blob/83962086337d77bf0a81c76890132115286503e3/assets/sass/vendor/_mdc-checkbox.scss#L35-L41

nfmohit replied:
Whoops, completely forgot about this one, updated, thank you for the reminder!

aaemnnosttv commented on assets/sass/components/key-metrics/_googlesitekit-km-selection-panel.scss:
```diff
@@ -0,0 +1,132 @@
+
+		.googlesitekit-km-selection-panel-footer__note {
+			color: $c-surfaces-on-surface;
+			margin: 1.15em 0 0;
```
This seems like an arbitrary value, where does it come from and do we have a better variable to use instead?

nfmohit replied:
The Figma design appears to apply a spacing of 14px here, and it looks like we do not have a good variable to apply that. For now, I've changed this to a raw `14px` value similar to other places we've done previously, but please let me know if we should do it in a different way, thanks!

aaemnnosttv reviewed (commented):
Thanks @nfmohit ! This one is nearly there, just a few things left to address.

nfmohit commented on assets/js/components/KeyMetrics/MetricsSelectionPanel/Header.js:
```diff
@@ -0,0 +1,82 @@
+					sprintf(
+						/* translators: %d: Number of selected metrics. */
+						__(
+							'<strong>%d of 4</strong> metrics selected',
```
Updated, thanks!

nfmohit commented on assets/js/components/KeyMetrics/MetricsSelectionPanel/MetricItem.js:
```diff
@@ -0,0 +1,105 @@
+						role="button"
+					>
+						<Checkbox
+							checked={ selectedMetrics?.includes( slug ) }
```
Updated, thank you!

nfmohit commented on assets/js/components/KeyMetrics/MetricsSelectionPanel/MetricItem.js:
```diff
@@ -0,0 +1,105 @@
+
+	const { setValues } = useDispatch( CORE_FORMS );
+
+	const onCheckboxChange = useCallback(
+		( event ) => {
+			setValues( KEY_METRICS_SELECTION_FORM, {
+				[ KEY_METRICS_SELECTED ]: event.target.checked
+					? selectedMetrics.concat( [ slug ] )
+					: selectedMetrics.filter(
+							( selectedMetric ) => selectedMetric !== slug
+					  ),
+			} );
+		},
+		[ selectedMetrics, setValues, slug ]
+	);
```
Updated, thanks!

nfmohit commented on assets/js/components/KeyMetrics/MetricsSelectionPanel/Metrics.js:
```diff
@@ -0,0 +1,68 @@
+export default function Metrics() {
+	const availableMetrics = useSelect( ( select ) =>
+		pickBy( KEY_METRICS_WIDGETS, ( _value, key ) => {
+			const widget = select( CORE_WIDGETS ).getWidget( key );
+
+			if ( ! widget ) {
+				return false;
+			}
+
+			return widget.modules.every( ( module ) =>
+				select( CORE_MODULES ).isModuleConnected( module )
```
Updated, thank you!

nfmohit commented on assets/js/components/KeyMetrics/MetricsSelectionPanel/index.test.js:
```diff
@@ -0,0 +1,191 @@
+			render( <MetricsSelectionPanel />, { registry } );
+
+			expect(
+				document.querySelector(
+					'.googlesitekit-km-selection-panel-metrics input[type=checkbox]:not(:checked)'
```
Updated, thank you!

nfmohit commented:
> Thanks @nfmohit ! This one is nearly there, just a few things left to address.

This is back to you for another look, thank you @aaemnnosttv!

aaemnnosttv commented on assets/js/components/KeyMetrics/MetricsSelectionPanel/Metrics.js:
```diff
@@ -32,19 +32,21 @@ import MetricItem from './MetricItem';
+		const { isModuleConnected } = select( CORE_MODULES );
+
+		return pickBy( KEY_METRICS_WIDGETS, ( _value, key ) => {
 			const widget = select( CORE_WIDGETS ).getWidget( key );
```
The same optimization we did with `isModuleConnected` applies here as well.

nfmohit replied:
Ah, excellent point, thank you for spotting this!

aaemnnosttv reviewed (commented):
Thanks @nfmohit ! There are a few final tweaks which are very trivial which I'll push now. Great work!

nfmohit commented on assets/js/components/KeyMetrics/MetricsSelectionPanel/index.js:
```diff
@@ -43,16 +45,37 @@ export default function MetricsSelectionPanel() {
+		// the modules that they depend on are connected. This helps prevent metrics
+		// with disconnected dependencies appear as selected in the selection panel.
+		return metrics.filter( ( slug ) => {
+			const widget = select( CORE_WIDGETS ).getWidget( slug );
```
Thank you!

nfmohit commented on assets/js/components/SideSheet.js:
```diff
@@ -39,6 +39,7 @@ import { ESCAPE } from '@wordpress/keycodes';
  * Internal dependencies
  */
 import Portal from './Portal';
+import FocusTrap from 'focus-trap-react';
```
Oops, my bad. I let the editor automatically import it, but later forgot to sort it correctly. Thank you!
